OrbitVulkanLayer/VulkanLayerController.h

```diff
@@ -0,0 +1,460 @@
+        layer_create_info->u.pLayerInfo->pfnNextGetInstanceProcAddr;
+    PFN_vkGetDeviceProcAddr next_get_device_proc_addr_function =
+        layer_create_info->u.pLayerInfo->pfnNextGetDeviceProcAddr;
+
+    // Advance linkage for next layer
+    layer_create_info->u.pLayerInfo = layer_create_info->u.pLayerInfo->pNext;
+
+    // Need to call vkCreateInstance down the chain to actually create the
+    // instance, as we need it to be alive in the create instance dispatch table.
+    auto create_device_function = absl::bit_cast<PFN_vkCreateDevice>(
+        next_get_instance_proc_addr_function(VK_NULL_HANDLE, "vkCreateDevice"));
+    VkResult result = create_device_function(physical_device, create_info, allocator, device);
+
+    dispatch_table_.CreateDeviceDispatchTable(*device, next_get_device_proc_addr_function);
+
+    device_manager_.TrackLogicalDevice(physical_device, *device);
+    timer_query_pool_.InitializeTimerQueryPool(*device);
+
+    return result;
+  }
+
+  [[nodiscard]] PFN_vkVoidFunction OnGetDeviceProcAddr(VkDevice device, const char* name) {
+    return dispatch_table_.GetDeviceProcAddr(device)(device, name);
+  }
+
+  [[nodiscard]] PFN_vkVoidFunction OnGetInstanceProcAddr(VkInstance instance, const char* name) {
+    return dispatch_table_.GetInstanceProcAddr(instance)(instance, name);
+  }
+
+  void OnDestroyInstance(VkInstance instance, const VkAllocationCallbacks* allocator) {
+    PFN_vkDestroyInstance destroy_instance_function = dispatch_table_.DestroyInstance(instance);
+    CHECK(destroy_instance_function != nullptr);
+    dispatch_table_.RemoveInstanceDispatchTable(instance);
+
+    destroy_instance_function(instance, allocator);
+
+    CloseVulkanLayerProducerIfNecessary();
+  }
+
+  void OnDestroyDevice(VkDevice device, const VkAllocationCallbacks* allocator) {
+    PFN_vkDestroyDevice destroy_device_function = dispatch_table_.DestroyDevice(device);
+    CHECK(destroy_device_function != nullptr);
+    device_manager_.UntrackLogicalDevice(device);
+    dispatch_table_.RemoveDeviceDispatchTable(device);
+
+    destroy_device_function(device, allocator);
+  }
+
+  // ----------------------------------------------------------------------------
+  // Core layer logic
+  // ----------------------------------------------------------------------------
+
+  [[nodiscard]] VkResult OnResetCommandPool(VkDevice device, VkCommandPool command_pool,
+                                            VkCommandPoolResetFlags flags) {
+    VkResult result = dispatch_table_.ResetCommandPool(device)(device, command_pool, flags);
+    submission_tracker_.ResetCommandPool(command_pool);
+    return result;
+  }
+
+  [[nodiscard]] VkResult OnAllocateCommandBuffers(VkDevice device,
+                                                  const VkCommandBufferAllocateInfo* allocate_info,
+                                                  VkCommandBuffer* command_buffers) {
+    VkResult result =
```

Comment:
Check result before proceeding? It seems if allocation here fails, we should not do the next steps..

Reply:
As above, I added an `if` rather then an `CHECK`.

Reply:
Looks good.